### DISCUSSION TITLE
Traversing the ryml::Tree

### DISCUSSION THREAD

**Discussion by rizwankadhar** [biojppm/rapidyaml#245] [Q&A] [answered]
Hi,
I want to traverse a ryml::Tree and create a std::vector<keyData> out of it. Where keyData will store every key-data pair from the tree with their parent, if any.

```c++
struct keyData {
    ryml::csubstr key;
    ryml::csubstr parent;
    ryml::csubstr data;
    keyData() {}
    keyData(ryml::csubstr k, ryml::csubstr p, ryml::csubstr d) : key(k), parent(p), data(d) {}
};
```

For example, if I have a tree like:
```c++
ryml::Tree tree = ryml::parse_in_arena(R"(doe: "a deer, a female deer"
ray: "a drop of golden sun"
pi: 3.14159
xmas: true
french-hens: 3
calling-birds:
   - huey
   - dewey
   - louie
   - fred
xmas-fifth-day:
  calling-birds: four
  french-hens: 3
  golden-rings: 5
  partridges:
    count: 1
    location: a pear tree
  turtle-doves: two
cars: GTO
)");
```
What I want as output is something like:
```c++
<{"doe", "", "a deer, a female deer"}, {"pi", "", "3.14159"} ... {"calling-birds", "", "[huey, dewey, louie, fred]"}, 
    {"xmas-fifth-day", "", "calling-birds: four
  french-hens: 3
  golden-rings: 5
  partridges:
    count: 1
    location: a pear tree
  turtle-doves: two"}, {"calling-birds", "xmas-fifth-day", "four"} ... {"count", "partridges", "1"} ... {"cars", "", "GTO"}>
```
So, basically every key has a parent, which can be null, a data, which can be a map or a sequence. And I want to go through all the nodes recursively.

I tried something like this, but it does not seem to be working properly, because for a map or a seq, I can not get their values or keys separately. This approach of mine may have other problems as well.

```c++
void getstr(ryml::NodeRef node, ryml::csubstr parent, vector<keyData> &vec)
{
    if (!node.has_children()) vec.push_back(keyData(node.has_key() ? node.key() : ryml::csubstr{}, parent, node.has_val() ? node.val() : ryml::csubstr{}));
    else {
        if (node.is_map())
        {
            std::string src = "";
            
            for (ryml::NodeRef ch : node[0].siblings())
            {
                src = src + (ch.has_val() ? ch.val().str : "");
            }
            ryml::csubstr data = ryml::to_substr(src);
            vec.push_back(keyData(node.has_key() ? node.key() : ryml::csubstr{}, parent, data));
        }
        else if (node.is_map())
        {
            std::string src = "";
            
            for (ryml::NodeRef ch : node[0].siblings())
            {
                src = src + (ch.has_key() ? ch.key().str : "") + (ch.has_val() ? ch.val().str : "");
            }
            ryml::csubstr data = ryml::to_substr(src);
            vec.push_back(keyData(node.has_key() ? node.key() : ryml::csubstr{}, parent, data));
        }
        for (ryml::NodeRef n : node.children())
        {
            getstr(n, node.has_key() ? node.key() : ryml::csubstr{}, vec);
        }
    }
```

I would highly appreciate if you can tell me of way to get this required output from a ryml::Tree constructed out of Yaml file.

Thanks a lot.

**Comment by rizwankadhar** [comment]
@biojppm Should I assume its not possible to get the nested key-value pairs?

**Comment by biojppm** [comment]
You can do anything, but the question is whether it is worth it.

I'm not sure what you're trying to achieve by creating this vector, or if I can help you achieve it; the problem seems very specific.

But I notice that for example, creating `{"calling-birds", "", "[huey, dewey, louie, fred]"}` will require emitting `tree["calling birds"]` to create the string `"[huey, dewey, louie, fred]"`. You are doing a hand-rolled emitter which will fail to descend into nested nodes, eg in `tree["xmas-fifth-day"]`. Use instead the `ryml::emitrs<std::string>(node)` function to get the string.

There are other problems in your code:
  * you have twice `if(node.is_map())`. one of them must be `if(node.is_seq())`
  * `node[0].siblings()` is exactly the same as `node.children()`

**Comment by biojppm** [comment, marked as answer]
Also I noticed that you're storing the `csubstr` view into a temporary `std::string`. That's sure to cause a read-after-free crash.

Do something like this:
```c++
struct keyData {
    std::string key; // store as std::string!!!
    std::string parent;
    std::string data;
    keyData() {}
    keyData(ryml::csubstr k, ryml::csubstr p, ryml::csubstr d) : key(k.str, k.len), parent(p.str, p.len), data(d.str, d.len) {}
};
void getstr(ryml::NodeRef node, ryml::csubstr parent, vector<keyData> &vec)
{
    auto getkey = [](NodeRef node){ return node.has_key() ? node.key() : ryml::csubstr{}; };
    auto getval = [](NodeRef node){ return node.has_val() ? node.val() : ryml::csubstr{}; };
    if (!node.has_children())
        vec.push_back(keyData(getkey(node), parent, getval(node)));
    else {
        std::string src = ryml::emitrs<std::string>(node);
        vec.push_back(keyData(getkey(node), parent, ryml::to_csubstr(src)));
        for (ryml::NodeRef n : node.children())
        {
            getstr(n, getkey(n), vec);
        }
    }
```
But maybe you will also have to do something about the key on the result of `emitrs()`.

**Reply by biojppm** [reply]
@rizwankadhar did this help? If so, please mark this as answered.

**Reply by rizwankadhar** [reply]
Yes, this helps. Thanks a a lot!

Now, I will try to figure out how to remove the key from the result of the emitrs and get the list for a list and a map for a map from that string.

**Reply by biojppm** [reply]
There's something else. The first if should be `if(!node.is_container())` instead of `if(!node.has_children())`. The latter will match empty containers, which is not wanted.
